File: week-2/problem-set/prob10.py
"""Prob10 example.

This file belongs to the week 2 data-structure and graph practice solutions. It uses tower_of_hanoi to illustrate the main idea with a small runnable example.
"""

def tower_of_hanoi(n, src, dest, aux):
    """Solve the Tower of Hanoi problem iteratively using an explicit stack.

    Args:
        n (int): The n input used by this function.
        src (str): The src input used by this function.
        dest (str): The dest input used by this function.
        aux (str): The aux input used by this function.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    stack = []
    stack.append((n, src, dest, aux))

    while stack:
        n, src, dest, aux = stack.pop()
        if n == 1:
            print(f"Move disk from {src} to {dest}")
        else:
            stack.append((n-1, aux, dest, src))
            stack.append((1, src, dest, aux))
            stack.append((n-1, src, aux, dest))

# Example usage:
tower_of_hanoi(3, 'A', 'C', 'B')

File: week-2/problem-set/prob10.cpp
/**
 * @file prob10.cpp
 * @brief C++ version of the iterative Tower of Hanoi example.
 */

#include <iostream>
#include <stack>
#include <tuple>

using namespace std;

/**
 * @brief Solve the Tower of Hanoi problem iteratively with an explicit stack.
 * @param n int Number of disks in the puzzle.
 * @param src char Name of the source peg.
 * @param dest char Name of the destination peg.
 * @param aux char Name of the auxiliary peg.
 * @return void This function prints each move and does not return a value.
 */
void tower_of_hanoi(int n, char src, char dest, char aux) {
    stack<tuple<int, char, char, char>> work;
    work.push({n, src, dest, aux});

    while (!work.empty()) {
        auto [disks, source, destination, helper] = work.top();
        work.pop();

        if (disks == 1) {
            cout << "Move disk from " << source << " to " << destination << '\n';
        } else {
            work.push({disks - 1, helper, destination, source});
            work.push({1, source, destination, helper});
            work.push({disks - 1, source, helper, destination});
        }
    }
}

/**
 * @brief Run the iterative Tower of Hanoi demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    tower_of_hanoi(3, 'A', 'C', 'B');
    return 0;
}
